File: Hard/1964. Find the Longest Valid Obstacle Course at Each Position/Solution.cpp
class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& obstacles) {
        int n = (int) obstacles.size();
        vector<int> dp, ret(n);

        for(int i = 0; i < n; i++){
            if(dp.empty() or dp.back() <= obstacles[i]) {
                dp.emplace_back(obstacles[i]);
                ret[i] = (int)dp.size();
            } else {
                auto it = upper_bound(dp.begin(), dp.end(), obstacles[i]);
                *it = obstacles[i];
                ret[i] = it - dp.begin() + 1;
            }
        }

        return ret;
    }
};

File: Hard/1964. Find the Longest Valid Obstacle Course at Each Position/Solution.java
class Solution {
    public static int upperBound(int[] arr, int left, int right, int key) {
        int mid, result = -1;

        while (left <= right) {
            mid = left + (right - left) / 2;

            if (arr[mid] <= key) {
                left = mid + 1;
            } else {
                result = mid;
                right = mid - 1;
            }
        }

        return result;
    }

    public int[] longestObstacleCourseAtEachPosition(int[] obstacles) {
        int n = obstacles.length;
        int[] dp = new int[n];
        int[] ret = new int[n];

        int len = 0;
        for(int i = 0; i < n; i++){
            if(len == 0 || dp[len - 1] <= obstacles[i]){
                dp[len++] = obstacles[i];
                ret[i] = len;
            } else {
                int idx = upperBound(dp, 0, len - 1, obstacles[i]);
                dp[idx] = obstacles[i];
                ret[i] = idx + 1;
            }
        }

        return ret;
    }
}